pkgs/development/compilers/go/1.11.nix

```diff
@@ -132,27 +138,35 @@ stdenv.mkDerivation rec {
   '';
 
   GOOS = stdenv.hostPlatform.parsed.kernel.name;
-  GOARCH = {
-    "i686" = "386";
-    "x86_64" = "amd64";
-    "aarch64" = "arm64";
-    "arm" = "arm";
-    "armv5tel" = "arm";
-    "armv6l" = "arm";
-    "armv7l" = "arm";
-  }.${stdenv.hostPlatform.parsed.cpu.name} or (throw "Unsupported system");
+  GOARCH = goarch stdenv.hostPlatform;
+  GOHOSTOS = stdenv.buildPlatform.parsed.kernel.name;
+  GOHOSTARCH = goarch stdenv.buildPlatform;
+
   GOARM = toString (stdenv.lib.intersectLists [(stdenv.hostPlatform.parsed.cpu.version or "")] ["5" "6" "7"]);
   GO386 = 387; # from Arch: don't assume sse2 on i686
   CGO_ENABLED = 1;
-  GOROOT_BOOTSTRAP = "${goBootstrap}/share/go";
   # Hopefully avoids test timeouts on Hydra
   GO_TEST_TIMEOUT_SCALE = 3;
 
-  # The go build actually checks for CC=*/clang and does something different, so we don't
-  # just want the generic `cc` here.
-  CC = if stdenv.isDarwin then "clang" else "cc";
+  # {CC,CXX}_FOR_TARGET must be only set for cross compilation case as go expect those
+  # to be different from CC/CXX
+  CC_FOR_TARGET = optionalString (stdenv.hostPlatform != stdenv.buildPlatform)
+    "${stdenv.cc.targetPrefix}cc";
+  CXX_FOR_TARGET = optionalString (stdenv.hostPlatform != stdenv.buildPlatform)
+    "${stdenv.cc.targetPrefix}c++";
 
   configurePhase = ''
+    runHook preConfigure
+    export GOCACHE=$TMPDIR/go-cache
+    # Points to build platform
+    export GOROOT_BOOTSTRAP=$(readlink -m $(command -v go)/../../share/go)
+    # The go build actually checks for CC=*/clang and does something different, so we don't
+    # just want the generic `cc` here.
+    CC=${if buildPackages.stdenv.isDarwin then
+        "${buildPackages.stdenv.cc}/bin/clang"
+      else
+        "${buildPackages.stdenv.cc}/bin/cc"}
```

Comment:
I had to move it here because the builder will override `CC` otherwise.

Reply:
It would be good to establish an interface for all the "cc" packages so we don't have to do that kind of case switching everywhere.

Here we assume that Darwin will always use clang and that all the other platforms will never use clang. I don't think this is necessarily true?

I don't think that we can ask all the "cc" packages to have a `bin/cc` because of the detection tools looking at the executable name. But we could export a `${cc}/.nix-support/set-cc` that `export CC=path/to/cc`. Or have a symlink that could be read. `export CC=$(readlink "${cc}/.nix-support/symlink-to-cc")`.

Reply:
I think this is bit out-of-scope of what this pull request tries to achieve.

Reply:
It might be worth a try, if we can actually remove the Darwin special case.

Reply:
https://github.com/NixOS/nixpkgs/pull/50914

Reply:
I removed it